File: remove-duplicates-from-sorted-list-ii.cc
class Solution {
public:
  ListNode *deleteDuplicates(ListNode *head) {
    ListNode* prev = NULL;
    ListNode* current = head;
    while (current != NULL) {
      if (prev != NULL && prev->val == current->val) {
        prev->next = current->next;
      } else {
          prev = current;
      }
      current = current->next;
    }
    return head;
  }
};

File: remove-duplicates-from-sorted-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode pre = new ListNode(0);
        ListNode tail = pre;
        while (head != null) {
            ListNode diff = head.next;
            while (diff != null && diff.val == head.val) diff = diff.next;
            if (diff == head.next) {
                tail.next = head;
                tail = head;
            }
            head = diff;
        }
        tail.next = null;
        return pre.next;
    }
}
